appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/OpenApiService.java

```diff
@@ -0,0 +1,159 @@
+import org.glassfish.api.event.Events;
+import org.glassfish.hk2.api.PostConstruct;
+import org.glassfish.hk2.api.PreDestroy;
+import org.glassfish.hk2.runlevel.RunLevel;
+import org.glassfish.internal.api.Globals;
+import org.glassfish.internal.data.ApplicationInfo;
+import org.glassfish.internal.deployment.Deployment;
+import org.glassfish.web.deployment.descriptor.WebBundleDescriptorImpl;
+import org.jvnet.hk2.annotations.Service;
+
+import fish.payara.microprofile.openapi.impl.admin.OpenApiServiceConfiguration;
+import fish.payara.microprofile.openapi.impl.config.OpenApiConfiguration;
+import fish.payara.microprofile.openapi.impl.model.OpenAPIImpl;
+import fish.payara.microprofile.openapi.impl.processor.ApplicationProcessor;
+import fish.payara.microprofile.openapi.impl.processor.BaseProcessor;
+import fish.payara.microprofile.openapi.impl.processor.FileProcessor;
+import fish.payara.microprofile.openapi.impl.processor.FilterProcessor;
+import fish.payara.microprofile.openapi.impl.processor.ModelReaderProcessor;
+
+@Service(name = "microprofile-openapi-service")
+@RunLevel(StartupRunLevel.VAL)
+public class OpenApiService implements PostConstruct, PreDestroy, EventListener {
+
+    private static final Logger LOGGER = Logger.getLogger(OpenApiService.class.getName());
+
+    private Deque<Map<ApplicationInfo, OpenAPI>> models;
+
+    @Inject
+    private Events events;
+
+    @Inject
+    private OpenApiServiceConfiguration config;
+
+    @Override
+    public void postConstruct() {
+        models = new ConcurrentLinkedDeque<>();
+        events.register(this);
+    }
+
+    @Override
+    public void preDestroy() {
+        events.unregister(this);
+    }
+
+    public boolean isEnabled() {
+        return Boolean.parseBoolean(config.getEnabled());
+    }
+
+    @Override
+    public void event(Event<?> event) {
+        if (event.is(Deployment.APPLICATION_STARTED)) {
+            // Get the application information
+            ApplicationInfo appInfo = (ApplicationInfo) event.hook();
+
+            // Create all the relevant resources
+            if (isEnabled()) {
+                OpenApiConfiguration appConfig = new OpenApiConfiguration(appInfo.getAppClassLoader());
+                Map<ApplicationInfo, OpenAPI> map = Collections.singletonMap(appInfo,
+                        createOpenApiDocument(appInfo.getAppClassLoader(), getContextRoot(appInfo), appConfig));
+                models.add(map);
+            }
+        } else if (event.is(Deployment.APPLICATION_UNLOADED)) {
+            ApplicationInfo appInfo = (ApplicationInfo) event.hook();
+            for (Map<ApplicationInfo, OpenAPI> map : models) {
+                if (map.keySet().toArray()[0].equals(appInfo)) {
+                    models.remove(map);
+                    break;
+                }
+            }
+        }
+    }
+
+    private String getContextRoot(ApplicationInfo appInfo) {
+        return appInfo.getMetaData(WebBundleDescriptorImpl.class).getContextRoot();
```

Comment:
You can get a NPE here for example when deploying an EJB jar as getMetaData I believe can return null.

Reply:
No worries, I've just made a fix for that.